### PR TITLE
Prepared statements cache computation improvements

### PR DESCRIPTION
**Pull request by bereng**
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

### CODE REVIEW

**Comment by bereng** [comment]
CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/1205/workflows/b3716eef-0dbe-4866-9417-08fb00d400cd) and [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/1205/workflows/dfc35f37-b0de-45e8-8c8f-86575540a636)

File: src/java/org/apache/cassandra/cql3/QueryProcessor.java [hunk 1]
```diff
@@ -149,11 +153,13 @@ public void preloadPreparedStatements()
                     clientState.setKeyspace(keyspace);
 
                 Prepared prepared = parseAndPrepare(query, clientState, false);
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/cql3/QueryProcessor.java]: `parseAndPrepare` might return `Prepared` with initialized size already so everything what will be needed after this line would be `preparedStatements.put(id, prepared)`

File: src/java/org/apache/cassandra/cql3/QueryProcessor.java [hunk 2]
```diff
@@ -149,11 +153,13 @@ public void preloadPreparedStatements()
...
-                preparedStatements.put(id, prepared);
+                int precomputedCacheEntrySize = measurePStatementCacheEntrySize(id, prepared);
+                Pair<Prepared, Integer> cacheValue = Pair.create(prepared, precomputedCacheEntrySize);
+                preparedStatements.put(id, cacheValue);
 
                 // Preload `null` statement for non-fully qualified statements, since it can't be parsed if loaded from cache and will be dropped
                 if (!prepared.fullyQualified)
-                    preparedStatements.get(computeId(query, null), (ignored_) -> prepared);
+                    preparedStatements.get(computeId(query, null), (ignored_) -> cacheValue);
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/cql3/QueryProcessor.java]: I don't understand what is that underscore in `ignored_` good for, but ... if we don't need `cacheValue` because of the above comment then we will basically revert this change.

Comment (bereng) [reply src/java/org/apache/cassandra/cql3/QueryProcessor.java]: This is some old style naming convention. Like using `_` for private members etc. Nothing to worry about.

File: src/java/org/apache/cassandra/cql3/QueryProcessor.java (hunk 1 above)
Comment (bereng) [reply src/java/org/apache/cassandra/cql3/QueryProcessor.java]: Yes. And with that the change is much less invasive.

**Comment by bereng** [comment]
Ci [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/1206/workflows/952965c2-005a-4425-a1c9-4c4d7cdce52e) and [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/1206/workflows/4ec073ca-7433-4bb2-b1f8-a4ec2663fa4c)

File: src/java/org/apache/cassandra/cql3/QueryHandler.java [hunk 3]
```diff
@@ -58,6 +58,9 @@ ResultMessage processBatch(BatchStatement statement,
 
     public static class Prepared
     {
+        // CASSANDRA-19986 Precomputed cache entry size might be available
+        public int cacheEntrySize = -1;
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/cql3/QueryHandler.java]: @bereng do you think this might have sense to make `volatile`?

File: src/java/org/apache/cassandra/cql3/QueryProcessor.java [hunk 4]
```diff
@@ -149,6 +149,7 @@ public void preloadPreparedStatements()
                     clientState.setKeyspace(keyspace);
 
                 Prepared prepared = parseAndPrepare(query, clientState, false);
+                prepared.cacheEntrySize = measureCacheEntrySize(id, prepared);
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/cql3/QueryProcessor.java]: @bereng what about putting `measureCacheEntrySize` into `prepareAndParse`? That method would be renamed to `measureSize` or similar. Are there any cases when we want to prepare and parse and _not_ have it measured? I think that by the act of preparing and parsing, it should be all measured already so we do not need to deal with it anymore. We most probably do not need to measure it again in `storePreparedStatement` (not sure about that though).

Comment (bereng) [reply src/java/org/apache/cassandra/cql3/QueryProcessor.java]: Yes I share your concern, I thought about that as well. But I decided I didn't want to burden pstmnts with the extra sizing in cases where it might not be needed. I couldn't come up with any concrete examples but any customer apps, side projects, other usages of code, some pstmnts generator stress tool, we don't know what happens in C* forks,  etc would be some. Wdyt?

File: src/java/org/apache/cassandra/cql3/QueryHandler.java (hunk 3 above)
Comment (bereng) [reply src/java/org/apache/cassandra/cql3/QueryHandler.java]: Right you are!

File: src/java/org/apache/cassandra/cql3/QueryProcessor.java (hunk 4 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/cql3/QueryProcessor.java]: perf tests need to compute it anyway, no? generator stress tool of prepared statements? never heard of that, is that a thing? Otherwise I think that measuring the size of it is rather innocent operation. If somebody ever complains (I really doubt that) we can always go back. We still need at least one compelling reason why not to do that.

Comment (bereng) [reply src/java/org/apache/cassandra/cql3/QueryProcessor.java]: I don't have a strong opinion here so :shrug: 

But when moving `measureCacheEntrySize(Object key, Prepared value)` I noticed it measures both the key and the value. Having a pstmnt calculate and hold it's size sgtm. But the pstmnt having it's size calculated with the size of a random key factored in makes no sense. Specially when that key is going to be implementation specific to a particular cache. That's not right.

What about having pstmnt having it's size precalculated and on `QueryProcessor` we pull that and calculate the cache entry size?

Comment (smiklosovic) [reply src/java/org/apache/cassandra/cql3/QueryProcessor.java]: by the way ... why do we need to compute the size with the key as well? We are caching a prepared statement so only its size should matter?

Key in that cache is MD5Digest and its size is same for every entry in that cache (if md5 hash is of same size for any input given).

Comment (bereng) [reply src/java/org/apache/cassandra/cql3/QueryProcessor.java]: That's how we calculate the cache size [here](https://github.com/apache/cassandra/pull/3605/files#diff-3ce9f6c4626262d6a4be26ecf3a57ef11fdb55537448def0c736115417487115R95). You need to measure both the key and value. Unlucky, otherwise yes it would have been nice to have the size in the pstmnt.

Comment (smiklosovic) [reply src/java/org/apache/cassandra/cql3/QueryProcessor.java]: javadoc on weigher() says:

_Specifies the weigher to use in determining the weight of entries. Entry weight is taken into consideration by maximumWeight(long) when determining which entries to evict, and use of this method requires a corresponding call to maximumWeight(long) prior to calling build. Weights are measured and recorded when entries are inserted into or updated in the cache, and are thus effectively static during the lifetime of a cache entry._

That is all true, but they think that key will / might be of different size. We have keys of a constant size because it is md5sum so we just dont need to take it into account, no?

however, `maximumWeight` docs says:

_Specifies the maximum weight of entries the cache may contain. Weight is determined using the Weigher specified with weigher, and use of this method requires a corresponding call to weigher prior to calling build.
Note that the cache may evict an entry before this limit is exceeded or temporarily exceed the threshold while evicting. As the cache size grows close to the maximum, the cache evicts entries that are less likely to be used again. For example, the cache may evict an entry because it hasn't been used recently or very often._

So I think that it will count on the size of the key upon eviction but can we not make that key of a constant size? What is MD5Digest producing? If the output is same everytime on its size then we can just put there a number as constant? 

In other words, if we do not include key into size as well, then this:

.maximumWeight(capacityToBytes(DatabaseDescriptor.getPreparedStatementsCacheSizeMiB()))

will not be true anymore, because the cache will contain both prepared statements _as well as_ all keys so their sizes _should_ be included, it is just the sizes of keys are effectively constants.

Comment (bereng) [reply src/java/org/apache/cassandra/cql3/QueryProcessor.java]: We spoke on Slack and it seems we agree having pstmnts hold the stmnt size alone is best, adding the key size later when needed.

File: src/java/org/apache/cassandra/cql3/QueryProcessor.java [hunk 5]
```diff
@@ -674,14 +678,14 @@ public static ResultMessage.Prepared storePreparedStatement(String queryString,
     {
         // Concatenate the current keyspace so we don't mix prepared statements between keyspace (#5352).
         // (if the keyspace is null, queryString has to have a fully-qualified keyspace so it's fine.
-        long statementSize = ObjectSizes.measureDeep(prepared.statement);
+        MD5Digest statementId = computeId(queryString, keyspace);
         // don't execute the statement if it's bigger than the allowed threshold
-        if (statementSize > capacityToBytes(DatabaseDescriptor.getPreparedStatementsCacheSizeMB()))
+        if ((prepared.pstmntSize + statementId.size()) > capacityToBytes(DatabaseDescriptor.getPreparedStatementsCacheSizeMB()))
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/cql3/QueryProcessor.java]: we might use  `getSizeForCache` for `(prepared.pstmntSize + statementId.size()`

File: src/java/org/apache/cassandra/cql3/QueryProcessor.java [hunk 6]
```diff
@@ -797,9 +801,17 @@ public static CQLStatement.Raw parseStatement(String queryStr) throws SyntaxExce
...
     }
 
-    private static int measure(Object key, Prepared value)
+    private static int measurePstmnt(Prepared value)
     {
-        return Ints.checkedCast(ObjectSizes.measureDeep(key) + ObjectSizes.measureDeep(value));
+        return Ints.checkedCast(ObjectSizes.measureDeep(value));
+    }
+
+    private static int getSizeForCache(MD5Digest key, Prepared value)
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/cql3/QueryProcessor.java]: @bereng I would rename this to `getSizeOfPreparedStatementForCache`

File: src/java/org/apache/cassandra/cql3/QueryProcessor.java [hunk 7]
```diff
@@ -797,9 +801,17 @@ public static CQLStatement.Raw parseStatement(String queryStr) throws SyntaxExce
...
+    {
+        if (value.pstmntSize < 0)
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/cql3/QueryProcessor.java:811]: @bereng is this actually ever going to happen? I think `assert value.pstmntSize > 0 : Precomputed prepared statement size not available` is enough. `> 0` because I don't think having size of 0 makes sense. It is invalid size too.

File: src/java/org/apache/cassandra/cql3/QueryProcessor.java (hunk 6 above)
Comment (bereng) [reply src/java/org/apache/cassandra/cql3/QueryProcessor.java]: sgtm

File: src/java/org/apache/cassandra/cql3/QueryProcessor.java (hunk 5 above)
Comment (bereng) [reply src/java/org/apache/cassandra/cql3/QueryProcessor.java]: +1

File: src/java/org/apache/cassandra/cql3/QueryProcessor.java (hunk 7 above)
Comment (bereng) [reply src/java/org/apache/cassandra/cql3/QueryProcessor.java:811]: Yes this is possible for yet non-parsed/prepared statements, DTO pstmnts, etc So I think I'll leave it as it is.

**Comment by bereng** [comment]
CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/1222/workflows/4fab084d-d59f-4d84-9c96-71c59f30e9fd) [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/1222/workflows/a2fe3d6e-05b3-4487-a4e4-abaee133c327)

LGTM where the only failure passes locally, flaky test imo.

**Comment by bereng** [comment]
[Here](https://app.circleci.com/pipelines/github/bereng/cassandra/1227/workflows/a19b1fb0-a983-455a-9b59-ff99327a8cdc/jobs/43535) are 500 repeats for the detected flaky test fix
